Geometry_Engine/Query/GeometryHash.cs

```diff
@@ -0,0 +1,446 @@
+        private static double[] GeometryHash(this Extrusion obj, double translationFactor)
+        {
+            translationFactor += (int)TypeTranslationFactor.Extrusion;
+
+            return obj.Curve.ITranslate(obj.Direction).GeometryHash(translationFactor)
+                .Concat(obj.Curve.GeometryHash(translationFactor)).ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for a Loft is calculated as the GeometryHash of its curves.")]
+        private static double[] GeometryHash(this Loft obj, double translationFactor)
+        {
+            translationFactor += (int)TypeTranslationFactor.Loft;
+
+            return obj.Curves.GeometryHash(translationFactor);
+        }
+
+        /***************************************************/
+
+        [Description("Moving control points by a translation factor composed by the weights " +
+          "and a subarray of the knot vector. " +
+          "The subarray is made by picking as many elements from the knot vector as the curve degree value.")]
+        private static double[] GeometryHash(this NurbsSurface obj, double translationFactor)
+        {
+            translationFactor += (int)TypeTranslationFactor.NurbsSurface;
+
+            List<int> uv = obj.UVCount();
+
+            List<double> uKnots = obj.UKnots.ToList();
+            List<double> vKnots = obj.VKnots.ToList();
+
+            List<double> concatenated = new List<double>();
+            for (int i = 0; i < uv[0]; i++)
+            {
+                double uSum = uKnots.GetRange(i, obj.UDegree).Sum();
+                for (int j = 0; j < uv[1]; j++)
+                {
+                    int ptIndex = i * uv[1] + j;
+                    double vSum = vKnots.GetRange(j, obj.VDegree).Sum();
+                    double[] doubles = obj.ControlPoints[ptIndex].GeometryHash(uSum + vSum + obj.Weights[ptIndex] + translationFactor);
+                    concatenated.AddRange(doubles);
+                }
+            }
+
+            return concatenated
+                .Concat(obj.InnerTrims.SelectMany(it => it.GeometryHash(translationFactor)))
+                .Concat(obj.OuterTrims.SelectMany(it => it.GeometryHash(translationFactor))).ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for a Pipe is calculated as the GeometryHash of its centreline translated by its radius," +
+            "then concatenated with the GeometryHash of its centreline's StartPoint for extra reliability.")]
+        private static double[] GeometryHash(this Pipe obj, double translationFactor)
+        {
+            translationFactor += (int)TypeTranslationFactor.Pipe;
+
+            double[] result = obj.Centreline.GeometryHash(translationFactor + obj.Radius);
+
+            if (obj.Capped)
+                result.Concat(obj.Centreline.StartPoint().GeometryHash(translationFactor + obj.Radius));
```

Comment:
Need to call interface method here. This just gets to the fallback method and reports an error.
```suggestion
                result.Concat(obj.Centreline.IStartPoint().GeometryHash(translationFactor + obj.Radius));
```